lib/IRGen/GenOpaque.cpp

```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
   Builder.CreateCall(stackRestoreFn, savedSP);
 }
 
+llvm::Value *IRGenFunction::emitStackAlloc(llvm::Value *size,
+                                           llvm::Value *align) {
+  auto returnType = IGM.Int8Ty;
+  auto returnPtrType = returnType->getPointerTo();
+  StackAddress stackPointer;
+  auto nullPointer = llvm::Constant::getNullValue(returnPtrType);
+
+  // Check that this function type can emit a stack allocation safely.
+  if (isAsync() || isCoroutine()) {
+    // These types of functions don't actually stack-allocate, and support for
+    // them is not needed at this time.
+    return nullPointer;
+  }
```

Comment:
Could you explain this behavior? Naively, I would expect that you should be able to use similar scratch-space like functionality in async functions too. There the allocation would be part of the async frame instead of the stack frame, but everything else should work as-is.

Maybe the primitive should be called "frameAlloc" or something so that it is more general?

Reply:
Async functions don't appear to emit `alloca` calls for dynamically-sized ranges. Instead, they heap-allocate and later free it. Since the API I'm proposing is not `async`, it doesn't need to care about these cases.

Reply:
Ah, right. I wrote this earlier but later realized that the size is not necessarily statically known.